kernel-rs/src/sysfile.rs

```diff
@@ -157,124 +168,143 @@ pub unsafe fn sys_unlink() -> usize {
-    let mut ip: *mut Inode = dirlookup(dp, &name, ptr::null_mut());
-    if !ip.is_null() {
-        (*dp).unlockput();
-        (*ip).lock();
-        if typ as i32 == T_FILE && ((*ip).typ as i32 == T_FILE || (*ip).typ as i32 == T_DEVICE) {
-            return ip;
+unsafe fn create(path: &Path, typ: i16, major: u16, minor: u16) -> Result<InodeGuard<'static>, ()> {
+    let (ptr, name) = ok_or!(path.nameiparent(), return Err(()));
+    let mut dp = InodeGuard {
+        guard: (*ptr).lock(),
+        ptr,
+    };
+    // TODO: use other Result related functions
+    if let Ok(ptr2) = dp.dirlookup(&name, ptr::null_mut()) {
+        dp.unlockput();
+        let ip = InodeGuard {
+            guard: (*ptr2).lock(),
+            ptr: ptr2,
+        };
+        if typ == T_FILE && (ip.guard.typ == T_FILE || ip.guard.typ == T_DEVICE) {
+            return Ok(ip);
         }
-        (*ip).unlockput();
-        return ptr::null_mut();
+        ip.unlockput();
+        return Err(());
     }
-    ip = Inode::alloc((*dp).dev, typ);
-    if ip.is_null() {
+    let ptr2 = Inode::alloc((*ptr).dev, typ);
+    if ptr2.is_null() {
         panic!("create: Inode::alloc");
     }
-    (*ip).lock();
-    (*ip).major = major;
-    (*ip).minor = minor;
-    (*ip).nlink = 1;
-    (*ip).update();
+    let mut ip = InodeGuard {
+        guard: (*ptr2).lock(),
+        ptr: ptr2,
+    };
+    ip.guard.major = major;
+    ip.guard.minor = minor;
+    ip.guard.nlink = 1;
+    ip.update();
 
     // Create . and .. entries.
-    if typ as i32 == T_DIR {
+    if typ == T_DIR {
         // for ".."
-        (*dp).nlink += 1;
-        (*dp).update();
+        dp.guard.nlink += 1;
+        dp.update();
 
         // No ip->nlink++ for ".": avoid cyclic ref count.
-        if dirlink(ip, FileName::from_bytes(b"."), (*ip).inum) < 0
-            || dirlink(ip, FileName::from_bytes(b".."), (*dp).inum) < 0
+        if !ip.dirlink(FileName::from_bytes(b"."), (*ptr2).inum)
+            || !ip.dirlink(FileName::from_bytes(b".."), (*ptr).inum)
         {
             panic!("create dots");
         }
     }
-    if dirlink(dp, &name, (*ip).inum) < 0 {
+    if !dp.dirlink(&name, (*ptr2).inum) {
         panic!("create: dirlink");
     }
-    (*dp).unlockput();
-    ip
+    dp.unlockput();
+    Ok(ip)
 }
 
 pub unsafe fn sys_open() -> usize {
     let mut path: [u8; MAXPATH] = [0; MAXPATH];
-    let ip: *mut Inode;
     let path = ok_or!(argstr(0, &mut path), return usize::MAX);
     let path = Path::new(path);
     let omode = ok_or!(argint(1), return usize::MAX);
     begin_op();
     let omode = FcntlFlags::from_bits_truncate(omode);
-    if omode.contains(FcntlFlags::O_CREATE) {
-        ip = create(path, T_FILE as i16, 0, 0);
-        if ip.is_null() {
+    let ip: InodeGuard<'_> = if omode.contains(FcntlFlags::O_CREATE) {
+        let ip = create(path, T_FILE, 0, 0);
+        if ip.is_err() {
             end_op();
             return usize::MAX;
         }
+        ip.unwrap()
```

Comment:
```suggestion
        ok_or!(create(path, T_FILE, 0, 0), {
            end_op();
            return usize::MAX;
        })
```
아래에도 unwrap 이 많은 데 unwrap 을 가능하면 피하는 것이 좋겠습니다.

Reply:
https://github.com/kaist-cp/rv6/pull/228/commits/c3041f1cef95a9ceeeabf130b986f5434871eda4